newrelic/newrelic-diagnostics-cli
# List available kafka brokers

# Issue
No issue, this a new feature/health check for Infrastructure agent/Kafka OHI

# Goals
Our goal is to add a new task for our Infrastructure agent that will look into the Kafka integration configuration and validate that the setting `zookeeper_path` (found in kafka-config.yml) is a valid path. If this path allows us to connect to Kafka server/brokers, then the path is valid.

# Implementation Details
1. I added a new go file under the category: Infra/Env/ValidateZookeeperPath
2. You'll notice this file has the linux extension: ValidateZookeeperPath_linux.go. This task was built and tested with only linux environment in mind since this is the most popular environment for new relic customers. We see very few windows tickets coming in at least for Infra OHI.
3. As I was mentioning above, connecting to the list of brokers is the way we troubleshoot that the zookeeper_path is valid. To connect to that list of brokers, we are following these steps:
- Verify we have found the kafka-config.yml where this path is set.
- Parse the yml to collect the keys and values for this setting.
- put together a command line using the values found. The cmd goes like this: `/path-to/zookeeper-shell.sh localhost:2181 ls valueforzookeeper_path/brokers/ids`
- Additionally, we need to find out the path to zookeeper-shell, a script that helps us run the mentioned command line. We can find its path by looking at some env vars: $PATH, $KAFKA_HOME, $ZOOKEEPER_HOME. Any of those may give us a hint where that script is located.
- With all that info collect, we can run the cmdline. If it returns an error, then we'll attempt to run the same cmdline minus their zookeeper_path set. It is possible that the path the user set is invalid and we'll verify this if the cmdline succeeds without using the path set.

# How to Test
If you are interested in smoke testing this task, then you can:
1. Checkout into this specific branch
2. Build the go binary for linux: `GOOS=linux go build`
3. Grab the binary for the root directory(newrelic-diagnostics-cli) and bring into your linux box where you'll be testing the binary.
4. You do not need to have a kafka server running, but at least you'll need to install the new relic infra agent + the kafka integration: https://docs.newrelic.com/docs/integrations/host-integrations/host-integrations-list/kafka-monitoring-integration
5. Then place the go binary inside the /etc/newrelic-infra/ directory. You run the ValidateZookeeperPath task with this cmd: `newrelic-diagnostics-cli -y -t Infra/Env/ValidateZookeeperPath`. You'll probably get a failure result, which is to be expected, because you do no have a kafka server
6. Now if you want to get fancy and install/configure a kafka server, then I recommend this tutorial: https://linuxhint.com/install-apache-kafka-ubuntu/
